File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_events/lib/src/state/events_state.dart
import '../models/events_models.dart';

enum EventsStatus { idle, loading, ready, error }

class EventsState {
  const EventsState({
    this.status = EventsStatus.idle,
    this.lastResult,
    this.error,
  });

  final EventsStatus status;
  final EventsPageResult<Object?>? lastResult;
  final String? error;

  EventsState loading() => const EventsState(status: EventsStatus.loading);

  EventsState loaded(EventsPageResult<Object?> result) =>
      EventsState(status: EventsStatus.ready, lastResult: result);

  EventsState failed(String message) =>
      EventsState(status: EventsStatus.error, error: message);
}

File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-events/src/state/eventsState.ts
import type { EventsPageResult } from "../types/eventsModels.js";

export interface EventsState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: EventsPageResult<unknown>;
  readonly error?: string;
}

export function initialEventsState(): EventsState {
  return { status: "idle" };
}

export function reduceEventsState(
  state: EventsState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: EventsPageResult<unknown> }
    | { type: "failed"; error: string },
): EventsState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}
